R/geocode.R

```diff
@@ -0,0 +1,101 @@
+#'   \code{"osm"}, \code{"census"}, \code{"arcgis"}, \code{"census_simple"}, \code{"geocodio"}, \code{"mapbox"},
+#'   \code{"google"}, \code{"bing"}, \code{"here"}, \code{"tomtom"}, \code{"nominatim"}, and \code{"tiger"}.
+#'
+#' @details This function attempts to geocode the specified location name using the chosen method. If the method requires an API key,
+#'   it will check for the necessary environment variable and return an error if the key is missing. The function outputs a data frame
+#'   containing the geocoded coordinates and location information.
+#'
+#' @return A data frame containing the following columns:
+#'   \item{name}{The name of the city or location.}
+#'   \item{country}{The country in which the location is situated, inferred from the display name.}
+#'   \item{lat}{The latitude of the location.}
+#'   \item{long}{The longitude of the location.}
+#'   \item{population}{The population of the location (not available in this function, returns \code{NA}).}
+#'
+#' @examples
+#' \dontrun{
+#'   # Geocode a location using OpenStreetMap (default)
+#'   geocode_raw("New York, USA")
+#'
+#'   # Geocode a location using Google (API key required)
+#'   geocode_raw("Tokyo, Japan", method = "google")
+#' }
+#'
+#' @export
+geocode <- function(location, method = "osm") {
+  if (missing(location) || !is.character(location)) {
+    stop("Please provide a valid location name as a string.")
+  }
+  
+  # Validate geocoding method
+  supported_methods <- c("osm", "census", "arcgis", "census_simple", "geocodio",
+                         "mapbox", "google", "bing", "here", "tomtom", "nominatim", "tiger")
+  if (!(method %in% supported_methods)) {
+    stop("Unsupported geocoding method. Please choose one of: ", paste(supported_methods, collapse = ", "))
+  }
+  
+  # Handle API keys for methods that require them
+  methods_with_keys <- c("google", "bing", "here", "tomtom", "mapbox", "geocodio")
+  if (method %in% methods_with_keys) {
+    api_key_env <- paste0(toupper(method), "_API_KEY")
+    api_key <- Sys.getenv(api_key_env)
+    if (api_key == "") {
+      stop(paste0("API key for ", method, " is required. Please set the '", api_key_env, "' environment variable."))
+    }
+  }
+  
+  result <- tryCatch({
+    geocode_df <- tibble::tibble(address = location) %>%
+      tidygeocoder::geocode(address, method = method, quiet = TRUE)
+    
+    if (any(is.na(geocode_df$lat)) || any(is.na(geocode_df$long))) {
+      stop("Geocoding failed: Unable to find coordinates for the provided location.")
+    }
+    
+    coords = list(
+      lat = geocode_df$lat[1],
+      lon = geocode_df$long[1],
+      display_name = location
+    )
+    
+    # Create a city-like df with geocoded coordinates
+    split_locations <- strsplit(coords$display_name, split = ",")[[1]]
+    
+    city <- data.frame(
+      name = split_locations[1],
+      country = split_locations[2],
+      lat = coords$lat,
+      long = coords$lon,
+      population = NA
```

Comment:
I don't think you need to make a column for `population`, it is mostly used by the twitter bot to filter out small cities.

Reply:
removed in my last commit